### PR TITLE
feat(hard_fork): Lightclient rollback

### PR DESCRIPTION
**Pull request by mtsitrin**
- Implement the OnHardFork hook in the LightClient module
- LightClient HardFork mode
  -  no LC updates are accepted until the fork is fully resolved.
- ResolveHardFork

### CODE REVIEW

File: x/lightclient/keeper/rollback.go [hunk 1]
```diff
@@ -2,24 +2,117 @@ package keeper
...
-		clientState, ok := k.ibcClientKeeper.GetClientState(ctx, clientId)
-		if !ok {
-			return errorsmod.Wrapf(types.ErrInvalidClientState, "client state for clientID %s not found", clientId)
-		}
+func (hook rollappHook) OnHardFork(ctx sdk.Context, rollappId string, height uint64) error {
+	hook.k.RollbackCanonicalClient(ctx, rollappId, height)
+	return nil
+}
+
+func (k Keeper) RollbackCanonicalClient(ctx sdk.Context, rollappId string, height uint64) {
```
Comment (omritoptix) [inline x/lightclient/keeper/rollback.go:19]: not clear if `height` is target height or fraud height. I suggest renaming it accordingly.

File: x/lightclient/keeper/rollback.go [hunk 2]
```diff
@@ -2,24 +2,117 @@ package keeper
...
 
-		tmClientState, ok := clientState.(*cometbfttypes.ClientState)
-		if !ok {
-			return errorsmod.Wrapf(types.ErrInvalidClientState, "client state with ID %s is not a tendermint client state", clientId)
+	// iterate over all consensus states and metadata in the client store
+	ibctm.IterateConsensusStateAscending(cs, func(h exported.Height) bool {
+		// if the height is lower than the target height, continue
+		if h.GetRevisionHeight() < height {
+			return false
 		}
```
Comment (omritoptix) [inline x/lightclient/keeper/rollback.go:32]: 2 things here:
1. if it's target height it should probably be  `<=` (unless it's the fraud height?)
2. the acennding iteration could be very pricy. desc iteration would be far more efficeint but I guess we don't have this api? thinking of 1 year of iteration.. not sure if the consensus states get pruned at some point?

File: x/lightclient/keeper/rollback.go [hunk 3]
```diff
@@ -2,24 +2,117 @@ package keeper
...
+
+	// unfreeze the client and set the latest height
+	tmClientState.FrozenHeight = clienttypes.ZeroHeight()
+	tmClientState.LatestHeight = clienttypes.NewHeight(1, height)
+
+	setClientState(clientStore, k.cdc, tmClientState)
+}
+
+// FIXME: assure there's no case with no proposer
+func (k Keeper) getValidatorHashForHeight(ctx sdk.Context, rollappId string, height uint64) []byte {
```
Comment (omritoptix) [inline x/lightclient/keeper/rollback.go:99]: `height` is not used

File: x/lightclient/keeper/rollback.go (hunk 2 above)
Comment (mtsitrin) [reply x/lightclient/keeper/rollback.go:32]: 1. it's ` fraud height`
2. consensus states got pruned after `TrustingPeriod`. anyway I'll change it to descending

**Pull request merged**